File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

bool checkSum(long creditNum);
int issuerCheck(long creditNum);

int main(void)
{
    //Prompts user for credit card number
    long credit = get_long("Number: ");
    //Runs Luhn’s algorithm and checks card issuer
    bool valid = checkSum(credit);
    int issuer = issuerCheck(credit);

    //Returns to user card type or invalid
    if (issuer == 1 && valid == true)
    {
        printf("VISA\n");
    }
    else if (issuer == 2 && valid == true)
    {
        printf("AMEX\n");
    }
    else if (issuer == 3 && valid == true)
    {
       printf("MASTERCARD\n");
    }
    else
    {
       printf("INVALID\n");
    }

}


//runs check sum to validate credit card number
bool checkSum(long creditNum)
{
int pos = 0, sumEven = 0, sumOdd = 0, tempDigit = 0;
  //Runs Luhn’s algorithm
  while (creditNum != 0)
  {
      int indexDigit = creditNum % 10;

      //Even numbers; beginning from the end of creditNum
      if (pos % 2 == 0)
      {
          sumEven += indexDigit;
      }
      //Odd numbers; beginning from penultimate number in creditNum. Multiplied by 2
      else
      {
        tempDigit = indexDigit * 2;

       if (tempDigit > 9)
       {
           while (tempDigit != 0)
           {
               int temp = 0;
               temp = temp + (tempDigit % 10);
               sumOdd += temp;
               tempDigit = tempDigit / 10;
           }
       }
       else
       {
           sumOdd += tempDigit;
       }
      }
      creditNum = creditNum / 10;
      pos++;
  }
//Gets total sum and check that trailing digit is 0
int totalSum = sumOdd + sumEven;
if (totalSum % 10 == 0)
{
    return true;
}
else
{
    return false;
}
}

//Checks issuer of a credit card number
int issuerCheck(long creditNum)
{
    int numLength = 0;
    long tempCredit = creditNum;



    //Counts card number length
    while (creditNum != 0)
    {
        int temp = creditNum % 10;
        creditNum = creditNum / 10;
        numLength++;
    }

    //Find leading digit for each card type and length
    long exponent = numLength - 2;
    long lengthDivisor = powl(10, exponent);
    int leadingDigit = tempCredit / lengthDivisor;

    //Check constraints to determine card issuer
    if ((numLength == 13 || numLength == 16) && (leadingDigit >= 40 && leadingDigit <= 49))
    {
        //VISA
        return 1;
    }
    else if (numLength == 15 && (leadingDigit == 34 || leadingDigit == 37))
    {
        //AMEX
        return 2;
    }
    else if (numLength == 16 && (leadingDigit >= 51 && leadingDigit <= 55))
    {
        //MASTERCARD
        return 3;
    }
    else
    {
        //Invalid
        return 4;
    }
}

File: pset6/credit/credit.py
from cs50 import get_int, get_string
from math import trunc


def main():
    cardNum = get_int("Card number: ")
    cSum = checkSum(cardNum)
    issuer = cardIssuer(cardNum)

    if cSum == True and issuer == 1:
        print("VISA")
    elif cSum == True and issuer == 2:
        print("AMEX")
    elif cSum == True and issuer == 3:
        print("MASTERCARD")
    else:
        print("INVALID")


def checkSum(cardNum):
    checkSum = sumOdd = sumEven = pos = temp = 0
    # calculates sum of odd and even position numbers
    # beginning from end of card number
    while cardNum >= 1:
        currentDigit = trunc(cardNum % 10)
        if pos % 2 == 0:
            sumEven += currentDigit
        else:
            temp = currentDigit * 2
            if temp > 9:
                n = 0
                while temp >= 1:
                    n = trunc((temp % 10)) + n
                    temp = temp / 10
                sumOdd += n
            else:
                sumOdd += temp
        cardNum = cardNum / 10
        pos += 1

    # Checks final digit of checkSum. If 0 -> Valid card num
    checkSum = sumOdd + sumEven
    if checkSum % 10 == 0:
        return True
    else:
        return False


def cardIssuer(cardNum):
    numLen = leadingNum = 0
    temp = cardNum

    while temp != 0:
        temp = trunc(temp / 10)
        numLen += 1
    leadingNum = trunc(cardNum / (10 ** (numLen - 2)))

    if (numLen == 13 or numLen == 16) and (leadingNum >= 40 and leadingNum <= 49):
        # It's a VISA
        return 1
    elif (numLen == 15) and (leadingNum == 34 or leadingNum == 37):
        # It's a American Express
        return 2
    elif (numLen == 16) and (leadingNum >= 51 and leadingNum <= 55):
        # It's a Master Card
        return 3
    else:
        # It's invalid
        return 4


main()